apps/web/src/views/Farms/v4/components/PoolsPage.tsx

```diff
@@ -178,13 +90,28 @@ export const PoolsPage = () => {
   const [cursorVisible, setCursorVisible] = useState(NUMBER_OF_FARMS_VISIBLE)
   const [sortOrder, setSortOrder] = useState<ISortOrder>(SORT_ORDER.NULL)
   const [sortField, setSortField] = useState<keyof IDataType | null>(null)
+  const [isPoolListExtended, setIsPoolListExtended] = useState(false)
   // data source
-  const farmPools = useFarmPools()
+  const { loaded: fetchFarmListLoaded, data: farmList } = useFarmPools()
   const { extendPools, fetchPoolList, resetExtendPools } = useExtendPools()
+  const allTokenMap = useAllTokensByChainIds(allChainIds)
 
   const poolList = useMemo(
-    () => (!farmPools.loaded ? UNIVERSAL_FARMS : [...farmPools.data, ...extendPools]),
-    [farmPools, extendPools],
+    () =>
+      fetchFarmListLoaded && farmList.length
+        ? [
+            ...farmList,
+            ...(isPoolListExtended
+              ? extendPools
+              : extendPools.filter(
+                  (pool) =>
+                    // non farming list need to do a whitelist filter
+                    pool.token0.wrapped.address in allTokenMap[pool.chainId] &&
+                    pool.token1.wrapped.address in allTokenMap[pool.chainId],
+                )),
+          ]
```

Comment:
use `concat` for better performance

```
const filteredExtendPools = isPoolListExtended ? extendPools : extendPools.filter(...)
const result = farmList.concat(filteredExtendPools)
```

Reply:
yep, concat have 10x performance improvement for a simplify test, fixed